### PR TITLE
[WIP] Add generation id to uploads.json (#3073)

### PR DESCRIPTION
**Pull request by donatello**
## Description

Adding generation id to the uploads.json makes sure that we do not
read a stale version of the file.
- Since we always read before we write uploads.json, we can use a 64-bit
  counter as the generation id.
- Reading uploads.json now requires a read quorum to agree on the
  generation id in XL.
- The generation id is incremented whenever an uploadID is added to or
  removed from uploads.json.
- Fix a bug in FS where a read uploads.json error was ignored.
- Minor refactoring to make the code easier to follow.
## Motivation and Context

https://github.com/minio/minio/issues/3073
## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
## Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

File: cmd/object-multipart-common.go [hunk 1]
```diff
@@ -35,6 +35,7 @@ type uploadInfo struct {
 type uploadsV1 struct {
 	Version string       `json:"version"`   // Version of the current `uploads.json`
 	Format  string       `json:"format"`    // Format of the current `uploads.json`
+	GenID   int64        `json:"GenId"`     // A generation id to identify latest written file.
```
Comment (harshavardhana) [inline cmd/object-multipart-common.go:38]: Don't use GenID stick to the same style as "xl.json" .

File: cmd/object-multipart-common.go [hunk 2]
```diff
@@ -66,6 +69,16 @@ func (u uploadsV1) Index(uploadID string) int {
 	return -1
 }
 
+// RemoveUploadIDAtIndex - removes upload ID at given index. Also
+// increments Generation ID.
+func (u uploadsV1) RemoveUploadIDAtIndex(idx int) {
+	// we ignore invalid index
+	if len(u.Uploads) > idx {
+		u.Uploads = append(u.Uploads[:idx], u.Uploads[idx+1:]...)
+		u.GenID++
```
Comment (harshavardhana) [inline cmd/object-multipart-common.go:78]: We used this approach before for `xl.json` - time is more meaningful.

File: cmd/xl-v1-multipart-common.go [hunk 3]
```diff
@@ -87,22 +87,55 @@ func (xl xlObjects) updateUploadsJSON(bucket, object string, uploadsJSON uploads
...
+	genIDCount := make(map[int64]int)
+	for i := 0; i < len(uploadResp); i++ {
+		if errs[i] != nil {
 			continue
 		}
-		break
+		// update the seen generation ID counter
+		gID := uploadResp[i].GenID
+		genIDCount[gID]++
+		// is the current gen-id "quorum" seen?
```
Comment (harshavardhana) [inline cmd/xl-v1-multipart-common.go:131]: Look at how `xl.json` is read in this scenario we need to implement something in similar lines of `readAllXlMetadata()`

**Comment by donatello** [comment]
@krishnasrinivas Yes checked with harsha. The change is needed because the existing behaviour of reading the first available upload.json can lead to stale information (a previous write failure to a disk that overall succeeded in quorum, could be the version that is being read). A notion of quorum is required in reading uploads.json. Adding the generation id is a way to do that.

The backend format change is not serious here. We can without loss of (any extra) correctness, assume that existing uploads.json (written by a version of minio without this change) have a generation id of 0 and continue from there. Of course the patch is not yet complete as tests don't yet pass.

Regarding using time as a generation id: Suppose `t` is the timestamp in the existing uploads.json. It is possible (though somewhat unlikely) that a server may select and use the same timestamp `t` when updating uploads.json - this can happen due to time-skew in different servers. But this is a problem because if only some of the uploads.json files are updated, it is not possible to tell which is the more recent version. In the case of uploads.json we are able to circumvent this using an incrementing counter as a time independent logical clock because we have the property that uploads.json is always read before being written/updated - thus we always have the previous value of the counter and are able to increment it with no extra cost.

**Review by harshavardhana — COMMENTED** [review]
> Regarding using time as a generation id: Suppose t is the timestamp in the existing uploads.json. It is possible (though somewhat unlikely) that a server may select and use the same timestamp t when updating uploads.json - this can happen due to time-skew in different servers. But this is a problem because if only some of the uploads.json files are updated, it is not possible to tell which is the more recent version. In the case of uploads.json we are able to circumvent this using an incrementing counter as a time independent logical clock because we have the property that uploads.json is always read before being written/updated - thus we always have the previous value of the counter and are able to increment it with no extra cost.

This i already explained that GenID is not useful. Modtime is the right value even if they are skewed we are in quorum whatever is in quorum is always the right copy. 

We don't need to invent monotonic clock style behavior here.

**Comment by harshavardhana** [comment]
> The backend format change is not serious here. We can without loss of (any extra) correctness, assume that existing uploads.json (written by a version of minio without this change) have a generation id of 0 and continue from there. Of course the patch is not yet complete as tests don't yet pass.

This we need to confirm first with AB as well.  This patch was to open up the discussion.

**Comment by donatello** [comment]
After discussing with AB, this is going to be tackled with a new algorithm. I will open a new PR as it is significantly different from this approach.

**Pull request closed without merging**